### PR TITLE
new module gatk/variants2table

### PR DESCRIPTION
**Pull request by lindenb**
This modules implements gatk/variants2table ( was https://nfcore.slack.com/archives/CJRH30T6V/p1725906645564699 )

- [X] This comment contains a description of changes (with reason).
- [X] If you've fixed a bug or added code that should be tested, add tests!
- [X] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [X] If necessary, include test data in your PR.
- [X] Remove all TODO statements.
- [X] Emit the `versions.yml` file.
- [X] Follow the naming conventions.
- [X] Follow the parameters requirements.
- [X] Follow the input/output options guidelines.
- [X] Add a resource `label`
- [X] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - For modules:
    - [ ] `nf-core modules test <MODULE> --profile docker`
    - [X] `nf-core modules test <MODULE> --profile singularity`
    - [X] `nf-core modules test <MODULE> --profile conda`

### CODE REVIEW

File: modules/nf-core/gatk4/variantstotable/meta.yml [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        Groovy Map containing reference information
+        e.g. [ id:'genome' ]
+  - dict:
+      type: file
+      description: Sequence dictionary of fastea file
+      pattern: "*.dict"
+output:
+  - table:
+      type: file
+      description: GATK output
```
Comment (fellen31) [inline modules/nf-core/gatk4/variantstotable/meta.yml:62]:
```suggestion
      description: Tab-delimited table of extracted fields from VCF
```

File: modules/nf-core/gatk4/variantstotable/tests/main.nf.test [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.dict', checkIfExists: true)
+                ]
+                """
+            }
+        }
+
+        then {
+            assertAll(
+                { assert process.success },
+                { assert snapshot(process.out).match()}
```
Comment (fellen31) [inline modules/nf-core/gatk4/variantstotable/tests/main.nf.test:41]:
```suggestion
                { assert snapshot(process.out).match() }
```

**Review by fellen31 — CHANGES_REQUESTED** [review]

File: modules/nf-core/gatk4/variantstotable/main.nf [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = meta.gatk_args ?: (task.ext.args ?: '')
+    def prefix = task.ext.prefix ?: "${meta.id}"
+
+    def avail_mem = 3072
+    if (!task.memory) {
+        log.info '[GATK VariantFiltration] Available memory not known - defaulting to 3GB. Specify process memory requirements to change this.'
```
[the remark this answers is not in the document]
Comment (lindenb) [reply modules/nf-core/gatk4/variantstotable/main.nf:29]: ah sorry, for all those copy-n-paste

File: modules/nf-core/gatk4/variantstotable/tests/main.nf.test [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    tag "gatk4"
+    tag "gatk4/variantstotable"
+
+    test("test1_gatk4_variant_to_table") {
+
+        when {
+            process {
+                """
+                input[0] = [
+                    [ id:'test' , gatk_args: '-F CHROM -F POS -F TYPE -GF AD' ], // meta map
```
[the remark this answers is not in the document]
Comment (lindenb) [reply modules/nf-core/gatk4/variantstotable/tests/main.nf.test:18]: while I know one can set the arguments in  a nextflow.config, I also like to be able to set the parameters in a channel to try multiple parameters for the same vcf:
```
ch = Channel.of(
   ' --split-multi-allelic -F CHROM -F POS -F TYPE -GF AD',
   ' -GF AD',
   '-F CHROM -F POS -F TYPE ' )
etc...
```

File: modules/nf-core/gatk4/variantstotable/tests/main.nf.test
[the remark this answers is not in the document]
Comment (lindenb) [reply modules/nf-core/gatk4/variantstotable/tests/main.nf.test:1]: @fellen31 what is a **stub** test ?

File: modules/nf-core/gatk4/variantstotable/tests/main.nf.test (hunk 4 above)
[the remark this answers is not in the document]
Comment (fellen31) [reply modules/nf-core/gatk4/variantstotable/tests/main.nf.test:18]: These arguments needs to be supplied via `ext.args` and not an input channel.

**Comment by lindenb** [comment]
@fellen31 back to you. Thanks for the review, I can remove this `meta.gatk_args` if needed but I like the flexibility of my implementation :-)

File: modules/nf-core/gatk4/variantstotable/tests/main.nf.test (hunk 4 above)
[the remark this answers is not in the document]
Comment (fellen31) [reply modules/nf-core/gatk4/variantstotable/tests/main.nf.test:18]: I see. I don't think any other module in nf-core implements this, you can check the 1.3 and 1.6 of the [module guidelines](https://nf-co.re/docs/guidelines/components/modules). 

You could: either use `nf-core modules patch` to change the module locally as you wish (after adding it to nf-core), or include the process 4 different times, e.g.:

```
include { GATK4_VARIANTSTOTABLE as VARIANTSTOTABLE_AF } from '...'
include { GATK4_VARIANTSTOTABLE as VARIANTSTOTABLE_GNOMAD_AF } from '...'
include { GATK4_VARIANTSTOTABLE as VARIANTSTOTABLE_EXOME_AF } from '...'
include { GATK4_VARIANTSTOTABLE as VARIANTSTOTABLE_UKBB_AF } from '...'
```
each with it's own `ext.args` in the config.

**Comment by cmatKhan** [comment]
Regarding a channel for input arguments -- it would be better to include the [--arguments_file](https://gatk.broadinstitute.org/hc/en-us/articles/360036896892-VariantsToTable#--arguments_file) as input. It technically should be anyway:

https://nf-co.re/docs/guidelines/components/modules#required-and-optional-input-files

> All mandatory and optional input files MUST be included in input channel definitions.

I'd also advocate including the [--intervals](https://gatk.broadinstitute.org/hc/en-us/articles/360036896892-VariantsToTable#--intervals) and [--exclude-intervals](https://gatk.broadinstitute.org/hc/en-us/articles/360036896892-VariantsToTable#--intervals) as file input. You would only set these when they are not null -- otherwise, the `-L` and `-XL` arguments can be passed in `task.args`

**Comment by lindenb** [comment]
Oh , I got this suggestion https://nfcore.slack.com/archives/CJRH30T6V/p1725975854237139?thread_ts=1725974540.609939&cid=CJRH30T6V

```
    ext.args = { meta.gatk_args }
```

I understand now, I'll update my code.

@cmatKhan I'll add both arguments.

File: modules/nf-core/gatk4/variantstotable/main.nf [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        VariantsToTable \\
+        ${args} \\
+        --variant "$vcf" \\
+        --output "${prefix}.tsv" \\
+        --reference $fasta \\
+        --tmp-dir . \\
+        $arguments_file_arg \\
+        $include_intervals \\
+        $exclude_intervals \\
+        $args
```
Comment (fellen31) [inline modules/nf-core/gatk4/variantstotable/main.nf:47]:
```suggestion
        VariantsToTable \\
        ${args} \\
        --variant $vcf \\
        --output ${prefix}.tsv \\
        --reference $fasta \\
        --tmp-dir . \\
        $arguments_file_arg \\
        $include_intervals \\
        $exclude_intervals
```

File: modules/nf-core/gatk4/variantstotable/main.nf [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        gatk4: \$(echo \$(gatk --version 2>&1) | sed 's/^.*(GATK) v//; s/ .*\$//')
+    END_VERSIONS
+    """
+    stub:
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    """
+    touch "${prefix}.tsv"
```
Comment (fellen31) [inline modules/nf-core/gatk4/variantstotable/main.nf:57]:
```suggestion
    touch ${prefix}.tsv
```

**Review by fellen31 — COMMENTED** [review]
NIce, thanks for removing the custom meta. There's two args in the module, could you just remove one of them?

**Review by fellen31 — APPROVED** [review]
👍

**Pull request merged**